### PR TITLE
inline values as decorators when debugging

### PR DESCRIPTION
**Pull request by nojvek**
1. [x] Write Tests.
2. [x] Figure out how to get words from lines. 
3. [x] Figure out how to stop flashing on stepping because onDidContinue passes in null stack frame everytime
4. [x] Figure out how to registerDecorator from public API. 
5. [x] setDecorators afterText doesn't work in python on lines with trailing ':'
6. [x] Merge as single commit
7. [x] Add screenshot

image

### CODE REVIEW

**Comment by nojvek** [comment]
@isidorn Please take a look at the first iteration.

**Comment by isidorn** [comment]
I will do a code review and provide feedback in the code a bit later today, let me first try to answer your questions
1. Awesome - might be a bit tricky though
2. Just checked and yes there is no nice API to get you all words on a line. As you figured out already you need to use a regex, however you should use a word regex for the particular language you are in. You should get the regex using this [method](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/modes/languageConfigurationRegistry.ts#L221). Here is a [sample usage](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/common/services/editorWorkerServiceImpl.ts#L349) 
3. Instead of using `editor.removeDecorations` and `editor.addDecorations` you need to use `editor.deltaDecorations` and probably keep track on your side of all the decorations from a previous state. DebugEditorModelManager is doing something similar now with breakpoints [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/browser/debugEditorModelManager.ts#L129) 
4. Why do you want to register a decorator from a public api? This feature will live in the vsccode core repo
5. No clue about that, we can investigate if it is a bug in the editor

**Comment by isidorn** [comment]
2. However if your name regex is more precise than the language word regex then we should probably use that

4. Here is a [public api](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L979) to set decorations in case you need it for some other use case

**Comment by isidorn** [comment]
@nojvek just to let you know that next week we have an endgame week - which means we take no new features. So it would be cool if we would try to land this feature together sometime this week. I have time on Thursday european time so I could try to wrap it up if something is still needed to be done then. Hope that works for you.

**Comment by nojvek** [comment]
Hi Isidor,

That's the plan. I'm going to send out a new iteration in a bit. Got some
bugs fixed. Fortunately I have Thursday and Friday off so can work on those
days full time.

**Comment by nojvek** [comment]
Hi Isidor,

I'm in need of using proper ES6 Map and Set. Can I change the typescript
tsconfig to include the es2015.collection lib ? The version of electron
vscode uses seems to have Set and Map defined so it should be fine right?

**Comment by isidorn** [comment]
Not sure if we can use it in the core
@jrieken is the right person to answer this

**Comment by jrieken** [comment]
re https://github.com/Microsoft/vscode/pull/16129#issuecomment-263522996 - Where do you need those? In an extension and the main-body of source code?

**Comment by isidorn** [comment]
@jrieken main-body of source code

**Comment by nojvek** [comment]
For time being I've added a map-set.d.ts in the typings folder.

The reason why I'm having hard time with using {} as map is that, I can't
quite set hasOwnProtype or prototype as keys, since they are already
defined properties. A scope variable could be this special object keys.

I've tested with python and javascript. Inline values work well so far.

Need to iron out a couple of bugs and clean up the registerDecorator call.

Its 3am here, I'm heading to bed.

Regards.

**Comment by jrieken** [comment]
> For time being I've added a map-set.d.ts in the typings folder.

That is not allowed :no_good_man: We have some technical debt there but in essence it would make ES6 Maps/Sets also available in the editor/base layer which we program against older runtime targets that don't support ES6. You might have a local version of those definitions, we have done that in a few [places](https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/workbench/api/node/mainThreadHeapService.ts#L16) but it is not nice. Also it seems you don't need that.

> The reason why I'm having hard time with using {} as map is that, I can't
quite set hasOwnProtype or prototype as keys, since they are already
defined properties. A scope variable could be this special object keys.

The fix for that is to *not* use `{}` but `Object.create(null)` or the utils defined in [`collections.ts`](https://github.com/Microsoft/vscode/blob/joh/xhr/src/vs/base/common/collections.ts#L23)

**Comment by nojvek** [comment]
I see. I'll undo the change and use collections helper. Thanks for the tip.

**Comment by nojvek** [comment]
Another question.

Having the inline values besides comments kind of looks weird. We match
against all words. Is it okay for me to getLineTokens for each line and
ignore the comment tokens?

**Comment by isidorn** [comment]
That sounds like a decent workaround for now. So +1
I believe tokenization is already dealing with words so it would be cool if you could just get tokens and not do any additional word magic on a line (which you are currently doing).

**Comment by nojvek** [comment]
Thanks.

Since getWordsRanges will mostly be creating a map from line tokens, makes
sense to move the  function to TextEditorModel and expose as a public right?

I wonder how I can keep track of when the editor model has been changed so
I can invalidate the cached wordsRangesMap

**Comment by isidorn** [comment]
@nojvek I would not move that to the TextEditorModel before I check with @alexandrudima. Will talk to him tomorrow in the office.

You can add a listener to the editor on changeModel. [Example](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts#L152)

**Comment by isidorn** [comment]
Ok I have discussed this with @alexandrudima and there are concerns
* The current implementation of putting decorations after text is suboptimal and can hurt performance
* Going through each word on a line can be very performant for very long lines
* What do we do for people that have word wrapping enabled. Via that setting they have specifically enabled lines not to overflow over some border. 

Due to the above, this feature will be under a setting which by default is false. Once we have no more performance concerns we can lift this flag. @nojvek you do not have to worry about introducing this flag, I can just put it on top once we polish this pr.

Also we need to put a limit on the line size and just do not compute anything for very long lines. Since currently we are matching the words on a line to a set of variables, we could also do it the other way around - match variables across a line. However this strategey does not work best for short lines which are the most common in practice. Thus I believe for starters we should just ignore long line.

As for the word wrap issue - that is something we need to think about. Might not require immediate action.

In a perfect world the debug extension would have a better ast understanding of the file - this knowledge would come from a language server but we currently do not have this. @weinand might know better if we plan to investigate into this any time in the future

Due to all of the above we do not have to rush this feature in for November. But let's see how it goes.

**Comment by nojvek** [comment]
1. Currently we only process max 100 variable scopes for perf
2. Every scope value is trimmed to max 50 chrs and ellipses added if too
long.
3. Maximum length of inline decorator will be no more than 150. All this
are constants that can easily be changed.
4. Adding this under feature switch makes sense. We can get good feedback
from beta users.
5. Not sure how to handle word wrapping but I believe that will be more of
a decorator implementation fix.
6. This might seem like a small feature but it makes a significant
difference in debugging experience. I'll do my best to get something
performant and usable.

**Comment by isidorn** [comment]
All those constraints make sense but we should introduce another one - the length of the editor line that is being processed. Because if the lenght of that line is quite large you will still go through each word and try to find it in the variables view, so some constraint needs to be added there.
We can handle word wrapping later, not for version 1.

fyi I am on vacation on Friday but I plan to code review, merge this feature and try to wrap it up on Monday.

**Comment by nojvek** [comment]
Sounds good. I'll skip tokenization if line length is greater than 1000
chars.

@alexandrudima <https://github.com/alexandrudima>, on getLineTokens there
is a inaccurateTokens parameter. What does that do?

Also what is the perf impact of getLineTokens for every line in a file?

**Comment by nojvek** [comment]
I've moved most of helper functions to debugInlineDecorators.ts where the core of logic really happens. Added 100% line, branch and function coverage for that file. 

I've tested on a bunch of different project types and it looks good so far. I've added a screenshot. 

Let me know if you need more changes. Its a nice sunny sunday, so I'll get back to this on Monday evening. Can't wait for this to get into nightly releases.

**Comment by isidorn** [comment]
@nojvek great work! And nice that you tested it in different projects!

I have added comments in the commit directly.

Let's move this feature out to the next release because there is no need for the rush and I would love the insiders to self host on this feature for a couple of weeks. This will also give us time to nicely adress the issues I brought up in the review.

**Comment by nojvek** [comment]
Hi Isidor, 

No problems. I am not seeing any comments in the review though. 

When do you think it will make insiders?

Regards.

**Comment by isidorn** [comment]
@nojvek ok that explains why you did not react on my previous comment from a couple of days ago.
Here is the direct link to my comments, you should be able to nicely see them
https://github.com/Microsoft/vscode/pull/16129/commits/d36a042fe25635191ab5f0c0bed987f060fda2cf

It should make insiders in about 8, 9 days when we wrap up our endgame.

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@nojvek my appologies - I did not click on submit review both times! I was used to using the old github UX which did not require an explict click on submit

**Comment by nojvek** [comment]
😳 still not able to see them.

May be paste the comments in conversation window.

Thanks.

**Comment by isidorn** [comment]
@nojvek you should be able to see them now, I made a mistake of not publishing them. Sorry for the misunderstanding!

File: src/vs/editor/browser/services/codeEditorServiceImpl.ts
Comment (nojvek) [inline src/vs/editor/browser/services/codeEditorServiceImpl.ts]: This was actually a VSCode bug that I discovered. You couldn't have after/before text decorators with a `'` in them. `_CSS_MAP.contentText` is defined as `contentText: 'content:\'{0}\';',` . This means the escaping of `"` actually did nothing. It should have been escaping single quotes `'`.  I'm fairly certain that this doesn't have side effects because its only escaping the contentText property and that's the only usage of it. 

Without this fix,python string values won't show up. Since python strings show up as `'cat bat mat'` in the watches view.

File: src/vs/editor/common/commonCodeEditor.ts
Comment (nojvek) [inline src/vs/editor/common/commonCodeEditor.ts]: setDecorators actually uses deltaDecorators. Inline decorators need to use after.contentText which ends up translating to a css:after block. For each of the unique decorators we need to create the selectors. By not using setDecorators we'll have to do that manually which will duplicate the logic in two places. 

setDecorators ends up calling `provider = new DecorationSubTypeOptionsProvider(this._styleSheet, key, parentTypeKey, options);` eventually.

setDecorators does the correct thing and provides the proper abstraction. Its just that before calling setDecorators, the decoratorTypeKey to be registered. That's why I added the registerDecoratorType api. 

I've spent hours trying to figure out if there was a nicer way to register the decoratorType. This seemed like the least obtrusive way since `this._registerDecorationType` was already being used.

The other workaround I can think is for setDecorators to auto-register the key if it doesn't exist.

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts
Comment (nojvek) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: Explained the issue about deltaDecorators on other thread.  We need the timeout anyway since setDecorators gives us a nice diff for free. We just need to debounce the null between the continues, otherwise we end up removing all decorators and creating them again rather than only changing the subset that actually are different.

File: src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts
Comment (nojvek) [inline src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts]: 👍

File: src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts
Comment (nojvek) [inline src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts]: I tried that. The issue is the tokens are whole line tokens since the `tokenizerSupport` is `null`  I tried using `model.setMode('javascript')` but that needs the javascript tokenizersupport to be registered and I couldn't fit all of the pieces to test that comments are ignored. So I wrote a simple stub that mocks `editor.getLineTokens()`.

Happy to change this if there is a way to plug in the javascript tokenizer.

File: src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts
Comment (nojvek) [inline src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts]: \b doesn't work since it doesn't consider `$` and `_`  as `word` character. Powershell, php would break seriously. Javascript allows $ and _ in variable names so they would have quirks as well.

File: src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts
Comment (nojvek) [inline src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts]: The generated parent classname with registerDecorationType looks like `ced-inlineDecorator-4` I could add the parent `:after` selector but it might look weird. 

I think once the decorator apis use an underlying implementation using spans rather than css :after selectors, we can nicely add the chrome like syntax highlighting for the inline values and move this to css. wdyt?

**Review by nojvek — COMMENTED** [review]
Fixed review feedback

**Comment by isidorn** [comment]
@nojvek thanks for your replies - we are in the middle of the endgame so I will reply to your comments in a couple of days once the bugfixing fuss is over. I also need to consult with @alexandrudima for the editor side of things concerning some of the questions you have raised.

**Comment by nojvek** [comment]
Sounds good.

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts
Comment (nojvek) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: It seems you commented on an older review. Does not apply to current the iteration .

**Comment by isidorn** [comment]
@nojvek feel free to ignore my comments from couple of days ago which I accidently only released yesterday

File: src/vs/editor/browser/services/codeEditorServiceImpl.ts
Comment (isidorn) [inline src/vs/editor/browser/services/codeEditorServiceImpl.ts]: Ok makes sense. But the best thing would be to provide a seperate PR that only fixes this issue.
That PR I will then assign to the editor people. As the description just copy this comment.

File: src/vs/editor/common/commonCodeEditor.ts
Comment (isidorn) [inline src/vs/editor/common/commonCodeEditor.ts]: Thanks for the explanation!
@aeschli how are you registering your folding end of line decorators? Can you provide a code pointer? Does it make sense to expose register editor decoration API as @nojvek is proposing here.

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts
Comment (isidorn) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: I understand that but this is still not a nice solution. Though it is fine for the first version and once we merge this in next week I can try to polish this if possible

File: src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts
Comment (isidorn) [inline src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts]: Ok makes sense. Though I hate introducing new regexes. Once we merge this in I can investigate if we can reuse some other word logic we have

File: src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts
Comment (isidorn) [inline src/vs/workbench/parts/debug/test/electron-browser/debugInlineDecorators.test.ts]: @aeschli do we have some mock tokenizer which we could reuse here?

File: src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts
Comment (isidorn) [inline src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts]: We would like to change the underlying implementation since it is already causing performance issues for folding. Though I will have to sync with @alexandrudima on this

**Comment by isidorn** [comment]
@nojvek added some comments. Seems like it is in a good state, I would lke to merge it in some time next week and then I can look into improving some of the issues we raised in the discussion (end of line decorations, using timeouts, creating a tokenizer for tests). Most of the team is still busy with the endgame so I will sync with them next week for input on improving implentation of decorations

**Comment by nojvek** [comment]
Thanks for the update @isidor

**Comment by nojvek** [comment]
@aeschli, In this instance managing the css on my own will probably be
almost the same as what setDecorators will do. The inline value decorators
are all after.contentText decorators which mean each of them will have a
corresponding ::after {..} css snippet.

That is why I wanted to add a registerDecoratorType api to the editor.

On Fri, Dec 9, 2016 at 8:48 AM, Martin Aeschlimann <notifications@github.com

**Comment by isidorn** [comment]
Hi @nojvek I was looking into mergin this in, the following would still be needed in order for me to do the merge

- [ ] As @aeschli mentioned [here](https://github.com/Microsoft/vscode/pull/16129#discussion_r91751609) no new editor api needs to be added. So please use existing API
- [ ] Changes in the `codeEditorServiceImpl.ts` need to be moved to a seperate PR
- [ ] merge your branch with latest from master and resolve conflicts

Other things I can polish up once we merge it in.

Thanks a lot!

**Comment by nojvek** [comment]
Sounds good. I'll get on it.

@aeschli, what did we decide regarding registerDecorator API ? It should
auto register key when calling setDecorators?

Regards.

**Comment by aeschli** [comment]
My suggestion is that you register your decoration type using ICodeEditorService.registerDecorationType

**Comment by isidorn** [comment]
ping @nojvek just to check if there are some updates here? Just fyi If you are busy I can also work on this.

File: src/vs/editor/browser/services/codeEditorServiceImpl.ts
Comment (nojvek) [inline src/vs/editor/browser/services/codeEditorServiceImpl.ts]: Migrated to separate PR. please see https://github.com/Microsoft/vscode/pull/18361/files

**Comment by nojvek** [comment]
@aeschli , codeEditorService is a private variable of codeEditor. It is not exposed. This is exactly what I am trying to do.  ICodeEditor.registerDecorationType ends up calling ICodeEditorService.registerDecorationType

Updated PR, as you recommend I am using ICodeEditorService.registerDecorationType `(<any>this.editor)._codeEditorService.registerDecorationType`. I get a feeling that you are very much against touching the code editor api.

I feel its worse solution than what we had before but gets the job done until vscode exposes a nicer api. I'm really not sure what you want me to do here. 😞

**Comment by aeschli** [comment]
@nojvek To get hold of the codeEditorService, use service injection in the class you are:
Add
`@ICodeEditorService codeEditorService: ICodeEditorService`
to the constructor of the DebugEditorContribution

**Comment by nojvek** [comment]
Whoa! Code injection dark magic.

Will give it a shot.

Is there any other major changes you need? I'd love to get this on this
week.

Will be on vacation until February.

Thanks.

**Comment by isidorn** [comment]
@nojvek just checked the PR and no other major changes are needed - except the code injection Martin mentioned. Once that is in I will merge this in.
Some polish work will be needed on top (adding setting, some refactoring, trying not to use timeouts) but that we will handle on our side.

Great work! :beers:

**Comment by nojvek** [comment]
Done. 

Just curious: Any chance someone can give me a brief overview of how vscode uses es6 decorators to make dependency injection work?

```
		@IDebugService private debugService: IDebugService,
		@IContextMenuService private contextMenuService: IContextMenuService,
		@IInstantiationService private instantiationService: IInstantiationService,
		@IContextKeyService contextKeyService: IContextKeyService,
		@ICommandService private commandService: ICommandService,
		@ICodeEditorService private codeEditorService: ICodeEditorService,
		@ITelemetryService private telemetryService: ITelemetryService
```

Am I correct to assume that the services are all singletons? i.e each of them is only instantiated once and then stored in a service registry somewhere ?

```
@editorContribution
export class DebugEditorContribution implements IDebugEditorContribution {
```

This is like adding the class instance to another registry as editor extensions/contributors ? In this sense if I create another class

```
@editorContribution
export class SyncEditorContribution implements IDebugEditorContribution {
```

vscode will instantiate my contribution, add to contributions registry and call its handler functions when editor change happens right? So with `@editorContribution` its possible to build functionality in vscode that syncs the editor with a peer or web browser for a collaborative interview?

**Comment by isidorn** [comment]
@nojvek thanks! I will merge it in today.
As for the code injection, yes every service is instantiated exactly once and more details can be found [here](https://github.com/Microsoft/vscode/wiki/Code-Organization#dependency-injection)

**Comment by isidorn** [comment]
@nojvek added setting, started using set & map (since it is now ok to do this - we no longer support IE10) and did some polish on top. I also plan to look into restructuring the debugInlineValues.ts so it is a class and not a bunch of helper methods. I am also considering not showing inline values for lines after the current line that is focussed.
6091d5d
24c9eba
db334b2

**Comment by nojvek** [comment]
Looks good. Thank you.

In the January release, This will be part of editor config right?

Can't wait to see this to completion.

**Comment by isidorn** [comment]
This can be set in settings > "debug.inlineValues": true

**Comment by luminaxster** [comment]
Wow!! Awesome work!
We are working on researching on and developing programming tools. What engine are you using to get the program state? Is it from scratch or using a VS API? Do you have some feedback(than can be shared) from the users that have used the feature? I highly recommend this [reading](http://www.cs.cmu.edu/~NatProg/papers/HCI%20in%20Tools%20Lifecycle%20-%20IEEE%20Comp%20PrePrint.pdf), if you haven't already, it goes "inline" with this feature.

**Comment by weinand** [comment]
@luminaxster since this feature made its first appearance in VS Code 1.9 (3 days ago) we do not yet have statistics. In addition this feature is not enabled by default because we are still working on it, so we do not expect widespread acceptance in this release.

**Pull request merged**
